janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/MultiPageEntryBuffer.java

```diff
@@ -0,0 +1,406 @@
+            }
+
+            //key not found, idx contains nearest index, negated
+            idx = (-idx - 1);
+
+            if (idx <= 0) //key is smaller than the first key in this page - search previous pages
+            {
+                high = mid - 1;
+            } else if (idx >= midPage.numEntries()) //key is bigger than the last in this page - search following pages
+            {
+                low = mid + 1;
+            } else //key was not found in this page, but the nearest key found is in this page
+            {
+                return mid;
+            }
+        }
+        return -(low + 1);  //key not found in any of the pages, and is outside the range of current pages - return the nearest page
+    }
+
+    @Override
+    public EntryList getSlice(KeySliceQuery query) {
+        if (pages == null || pages.size() == 0) {
+            return EntryList.EMPTY_LIST;
+        }
+
+        int startPageIdx = getPageIndex(query.getSliceStart());
+        if (startPageIdx < 0) {
+            startPageIdx = (-startPageIdx - 1);
+        }
+
+        int endPageLimit = getPageIndex(query.getSliceEnd());
+        if (endPageLimit < 0) {
+            endPageLimit = (-endPageLimit - 1);
+        } else {
+            endPageLimit++; //we want "insertion point" like in case when it wasn't found
+        }
+
+        if (startPageIdx < endPageLimit) {
+            int startIndex = pages.get(startPageIdx).getIndex(query.getSliceStart());
+            if (startIndex < 0) {
+                startIndex = (-startIndex - 1);
+            }
+
+            int endIndex = pages.get(endPageLimit - 1).getIndex(query.getSliceEnd());
+            if (endIndex < 0) {
+                endIndex = (-endIndex - 1);
+            }
+
+            MemoryEntryList result = new MemoryEntryList(0);
+            for (int i = startPageIdx; i < endPageLimit; i++) {
+                if (query.hasLimit() && result.size() >= query.getLimit()) {
+                    break;
+                }
+
+                BufferPage page = pages.get(i);
+                int start = (i == startPageIdx) ? startIndex : 0;
+                int end = (i == endPageLimit - 1) ? endIndex : page.numEntries();
+                for (int j = start; j < end; j++) {
+                    if (query.hasLimit() && result.size() >= query.getLimit()) {
+                        break;
+                    }
+
+                    //using getNoCopy here means the old page buffer will not be collected until the last of the Entries it used to hold is collected
+                    //however, since all Entries we return are managed by a single transaction and not presented to user code, this should not be a problem
+                    result.add(page.getNoCopy(j));
+                }
+            }
+
+            return result;
+        } else {
+            return EntryList.EMPTY_LIST;
+        }
+    }
+
+    @Override
+    public void mutate(Entry[] add, Entry[] del, int maxPageSize) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'mutate(Entry, Entry, int)' has an NPath complexity of 4476](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)